indexing-service/src/main/java/io/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTask.java

```diff
@@ -0,0 +1,645 @@
+                                            .withBasePersistDirectory(toolbox.getPersistDir());
+
+    logUnusedConfiguration(tuningConfig);
+
+    final FireDepartment fireDepartment = new FireDepartment(
+        dataSchema,
+        realtimeIOConfig,
+        tuningConfig
+    );
+
+    final RealtimeMetricsMonitor metricsMonitor = new RealtimeMetricsMonitor(
+        ImmutableList.of(fireDepartment),
+        ImmutableMap.of(
+            DruidMetrics.TASK_ID, new String[]{getId()}
+        )
+    );
+
+    this.metrics = fireDepartment.getMetrics();
+
+    Supplier<Committer> committerSupplier = null;
+    final File firehoseTempDir = toolbox.getFirehoseTemporaryDir();
+
+    DiscoveryDruidNode discoveryDruidNode = createDiscoveryDruidNode(toolbox);
+
+    appenderator = newAppenderator(dataSchema, tuningConfig, fireDepartment.getMetrics(), toolbox, lockingSegmentAnnouncer);
+    AppenderatorDriver driver = newDriver(dataSchema, appenderator, toolbox, fireDepartment.getMetrics());
+
+    try {
+      toolbox.getDataSegmentServerAnnouncer().announce();
+      toolbox.getDruidNodeAnnouncer().announce(discoveryDruidNode);
+
+      driver.startJob();
+
+      // Set up metrics emission
+      toolbox.getMonitorScheduler().addMonitor(metricsMonitor);
+
+      // Firehose temporary directory is automatically removed when this RealtimeIndexTask completes.
+      FileUtils.forceMkdir(firehoseTempDir);
+
+      // Delay firehose connection to avoid claiming input resources while the plumber is starting up.
+      final FirehoseFactory firehoseFactory = spec.getIOConfig().getFirehoseFactory();
+      final boolean firehoseDrainableByClosing = isFirehoseDrainableByClosing(firehoseFactory);
+
+      int sequenceNumber = 0;
+      String sequenceName = makeSequenceName(getId(), sequenceNumber);
+
+      final TransactionalSegmentPublisher publisher = (segments, commitMetadata) -> {
+        final SegmentTransactionalInsertAction action = new SegmentTransactionalInsertAction(segments);
+        return toolbox.getTaskActionClient().submit(action).isSuccess();
+      };
+
+      // Skip connecting firehose if we've been stopped before we got started.
+      synchronized (this) {
+        if (!gracefullyStopped) {
+          firehose = firehoseFactory.connect(spec.getDataSchema().getParser(), firehoseTempDir);
+          committerSupplier = Committers.supplierFromFirehose(firehose);
+        }
+      }
+
+      // Time to read data!
+      while (!gracefullyStopped && firehoseDrainableByClosing && firehose.hasMore()) {
+        try {
+          InputRow inputRow = firehose.nextRow();
+
+          if (inputRow == null) {
+            log.debug("Discarded null row, considering thrownAway.");
+            metrics.incrementThrownAway();
+          } else {
+            AppenderatorDriverAddResult addResult = driver.add(inputRow, sequenceName, committerSupplier);
+
+            if (addResult.isOk()) {
+              if (addResult.isPersistRequired()) {
```

Comment:
`addResult.isPsersistRequired()` always returns `false` because `driver.add(inputRow, sequenceName, committerSupplier)` is called which internally allowing automatic persisting.

Reply:
👍